support-frontend/assets/helpers/productPrice/__tests__/promotionsTests.js

```diff
@@ -1,28 +1,152 @@
+      description: 'example promotion2',
+      introductoryPrice: {
+        periodLength: 3,
+        periodType: 'issue',
+        price: 6.99,
+      },
+      name: 'examplePromo2',
+      promoCode: 5678,
+    });
+
+    expect(getAppliedPromo()).toEqual(null);
+  });
+});
+
+describe('applyDiscount', () => {
+  const productWithIntroductoryPrice = {
+    price: 12.99,
+    currency: 'EUR',
+    fixedTerm: false,
+    promotions: [{
+      name: 'Sept 2019 Discount',
+      description: '50% off for 3 months',
+      promoCode: 'GH86H9J',
+      introductoryPrice: { price: 6.99, periodLength: 6, periodType: 'issue' },
+    }],
+  };
+
+  const productWithDiscountedPrice = {
+    price: 12.99,
+    currency: 'EUR',
+    fixedTerm: false,
+    promotions: [{
+      name: 'Sept 2019 Discount',
+      description: '25% off',
+      promoCode: 'GH86H9J',
+      discountedPrice: 8.99,
+    }],
+  };
+
+  it('should return an updated price with a discount applied', () => {
+    expect(applyDiscount(productWithIntroductoryPrice, productWithIntroductoryPrice.promotions[0])).toEqual({
+      currency: 'EUR',
+      fixedTerm: false,
+      price: 6.99,
+      promotions: [{
+        name: 'Sept 2019 Discount',
+        description: '50% off for 3 months',
+        promoCode: 'GH86H9J',
+        introductoryPrice: { price: 6.99, periodLength: 6, periodType: 'issue' },
+      }],
+    });
+
+    expect(applyDiscount(productWithDiscountedPrice, productWithDiscountedPrice.promotions[0])).toEqual({
+      currency: 'EUR',
+      fixedTerm: false,
+      price: 8.99,
+      promotions: [{
+        name: 'Sept 2019 Discount',
+        description: '25% off',
+        promoCode: 'GH86H9J',
+        discountedPrice: 8.99,
+      }],
+    });
   });
+});
+
+describe('hasIntroductoryPrice', () => {
+  const promotionWithIntroductoryPrice = {
+    name: 'Sept 2019 Discount',
+    description: '50% off for 3 months',
+    promoCode: 'GH86H9J',
+    introductoryPrice: { price: 6.99, periodLength: 6, periodType: 'issue' },
+  };
 
+  const promotionWithoutIntroductoryPrice = {
+    name: 'Sept 2019 Discount',
+    description: '50% off for 3 months',
+    promoCode: 'GH86H9J',
+  };
+
+  it('should return true if there is an introductory price', () => {
+    expect(hasIntroductoryPrice(promotionWithIntroductoryPrice)).toEqual(true);
+    expect(hasIntroductoryPrice(promotionWithoutIntroductoryPrice)).toEqual(false);
+    expect(hasIntroductoryPrice()).toEqual(false);
+  });
+});
+
+describe('promotionHTML', () => {
+  it('should return promotion html if present', () => {
+    expect(promotionHTML()).toEqual(null);
+    expect(promotionHTML('Get 25% off').props).toEqual({ __EMOTION_TYPE_PLEASE_DO_NOT_USE__: 'span', css: '', dangerouslySetInnerHTML: { __html: 'Get 25% off' } });
+    expect(promotionHTML('Get 20% off', { tag: 'div' }).props).toEqual({ __EMOTION_TYPE_PLEASE_DO_NOT_USE__: 'div', css: '', dangerouslySetInnerHTML: { __html: 'Get 20% off' } });
```

Comment:
I don't understand a lot about client side tests, but is the `__EMOTION_TYPE_PLEASE_DO_NOT_USE__` something that has to be correct here?

Reply:
When we call the `promotionHTML` function the result from the shallow render is wrapped in a ForwardRef e.g. `<ForwardRef(render) __EMOTION_TYPE_PLEASE_DO_NOT_USE__="div" css="" dangerouslySetInnerHTML={{"__html": "Get 20% off"}} />` which I wasn't too sure how to get around

Though I could also update these tests to focus on the `dangerouslySetInnerHTML` object but thought as the function also allows you to specify which tag to use it might be most useful to see all of the props.